File: src/economy/weaverInputPolicy.ts
import {
  SPINNING_RETTING_FLAX_PER_CYCLE,
  SPINNING_RETTING_FLAX_WATER_PER_CYCLE,
  SPINNING_RETTING_WOOL_PER_CYCLE,
  WEAVER_LINEN_PER_CYCLE,
  WEAVER_YARN_PER_CYCLE,
} from '../generated/gameBalance.ts';
import type { BuildingState } from '../resources/types.ts';

export const WEAVER_INPUT_POLICY_AUTO = 0;
export const WEAVER_INPUT_POLICY_WOOL_FIRST = 1;
export const WEAVER_INPUT_POLICY_FLAX_FIRST = 2;

export type WeaverInputPolicy =
  | typeof WEAVER_INPUT_POLICY_AUTO
  | typeof WEAVER_INPUT_POLICY_WOOL_FIRST
  | typeof WEAVER_INPUT_POLICY_FLAX_FIRST;

export type WeaverFibreCommodity = 'wool' | 'flax' | 'yarn' | 'linen';

export const SPINNING_RETTING_INPUT_POLICY_PRESETS = [
  {
    policy: WEAVER_INPUT_POLICY_AUTO,
    label: 'Auto',
    hint: 'Shares raw-fibre carts and well service normally, using the route with the most complete cycles already staged.',
  },
  {
    policy: WEAVER_INPUT_POLICY_WOOL_FIRST,
    label: 'Yarn',
    hint: 'Spin wool into yarn only. Flax and water may still be stored for a later recipe change.',
  },
  {
    policy: WEAVER_INPUT_POLICY_FLAX_FIRST,
    label: 'Linen',
    hint: 'Ret flax into linen only. Wool may still be stored for a later recipe change.',
  },
] as const;

export const WEAVER_INPUT_POLICY_PRESETS = [
  {
    policy: WEAVER_INPUT_POLICY_AUTO,
    label: 'Auto',
    hint: 'Shares prepared-fibre carts normally, using the route with the most complete cycles already staged.',
  },
  {
    policy: WEAVER_INPUT_POLICY_WOOL_FIRST,
    label: 'Yarn clothing',
    hint: 'Weave clothing from yarn only. Linen may still be stored for a later recipe change.',
  },
  {
    policy: WEAVER_INPUT_POLICY_FLAX_FIRST,
    label: 'Linen clothing',
    hint: 'Weave clothing from linen only. Yarn may still be stored for a later recipe change.',
  },
] as const;

export function normalizeWeaverInputPolicy(
  policy: number | undefined,
): WeaverInputPolicy {
  if (policy === WEAVER_INPUT_POLICY_WOOL_FIRST) {
    return WEAVER_INPUT_POLICY_WOOL_FIRST;
  }
  if (policy === WEAVER_INPUT_POLICY_FLAX_FIRST) {
    return WEAVER_INPUT_POLICY_FLAX_FIRST;
  }
  return WEAVER_INPUT_POLICY_AUTO;
}

export function weaverInputPolicyLabel(policy: number | undefined): string {
  const normalized = normalizeWeaverInputPolicy(policy);
  return WEAVER_INPUT_POLICY_PRESETS.find((preset) => preset.policy === normalized)?.label
    ?? 'Auto';
}

export function spinningRettingInputPolicyLabel(policy: number | undefined): string {
  const normalized = normalizeWeaverInputPolicy(policy);
  return SPINNING_RETTING_INPUT_POLICY_PRESETS.find(
    (preset) => preset.policy === normalized,
  )?.label ?? 'Auto';
}

/**
 * Lower ranks win when equal-priority textile processors compete for one cart.
 * Policy route 1 means wool at the spinner and yarn at the Weaver; route 2
 * means flax at the spinner and linen at the Weaver. Auto remains neutral.
 */
export function weaverFibreDeliveryPreferenceRank(
  policy: number | undefined,
  commodity: WeaverFibreCommodity,
): number {
  const normalized = normalizeWeaverInputPolicy(policy);
  if (normalized === WEAVER_INPUT_POLICY_AUTO) return 1;
  const matches = commodity === 'wool' || commodity === 'yarn'
    ? normalized === WEAVER_INPUT_POLICY_WOOL_FIRST
    : normalized === WEAVER_INPUT_POLICY_FLAX_FIRST;
  return matches ? 0 : 2;
}

export function weaverFibreDeliveryPreferenceLabel(
  policy: number | undefined,
  commodity: WeaverFibreCommodity,
): string {
  switch (weaverFibreDeliveryPreferenceRank(policy, commodity)) {
    case 0:
      return `${commodity === 'wool' || commodity === 'flax'
        ? spinningRettingInputPolicyLabel(policy)
        : weaverInputPolicyLabel(policy)} match`;
    case 1:
      return 'Auto pool';
    default:
      return 'Fallback route';
  }
}

export function weaverUsesFlax(
  building: Pick<BuildingState, 'wool' | 'flax' | 'water' | 'weaverInputPolicy'>,
): boolean {
  const wool = Math.max(0, building.wool ?? 0);
  const flax = Math.max(0, building.flax ?? 0);
  const water = Math.max(0, building.water);
  const woolCycles = wool / Math.max(1e-6, SPINNING_RETTING_WOOL_PER_CYCLE);
  const flaxCycles = Math.min(
    flax / Math.max(1e-6, SPINNING_RETTING_FLAX_PER_CYCLE),
    water / Math.max(1e-6, SPINNING_RETTING_FLAX_WATER_PER_CYCLE),
  );
  switch (normalizeWeaverInputPolicy(building.weaverInputPolicy)) {
    case WEAVER_INPUT_POLICY_WOOL_FIRST:
      return false;
    case WEAVER_INPUT_POLICY_FLAX_FIRST:
      return true;
    case WEAVER_INPUT_POLICY_AUTO:
    default:
      return (flax > 1e-6 && wool <= 1e-6)
        || flaxCycles > woolCycles + 1e-9;
  }
}

export function weaverUsesLinen(
  building: Pick<BuildingState, 'yarn' | 'linen' | 'weaverInputPolicy'>,
): boolean {
  const yarn = Math.max(0, building.yarn ?? 0);
  const linen = Math.max(0, building.linen ?? 0);
  const yarnCycles = yarn / Math.max(1e-6, WEAVER_YARN_PER_CYCLE);
  const linenCycles = linen / Math.max(1e-6, WEAVER_LINEN_PER_CYCLE);
  switch (normalizeWeaverInputPolicy(building.weaverInputPolicy)) {
    case WEAVER_INPUT_POLICY_WOOL_FIRST:
      return false;
    case WEAVER_INPUT_POLICY_FLAX_FIRST:
      return true;
    case WEAVER_INPUT_POLICY_AUTO:
    default:
      return (linen > 1e-6 && yarn <= 1e-6)
        || linenCycles > yarnCycles + 1e-9;
  }
}

File: server/src/weaver_input_policy.rs
//! Persisted player-selected route preference for the two-stage textile chain.
//! The same stable codes select wool versus flax at the spinning/retting house
//! and yarn versus linen at the Weaver.

pub const WEAVER_INPUT_POLICY_AUTO: u8 = 0;
pub const WEAVER_INPUT_POLICY_WOOL_FIRST: u8 = 1;
pub const WEAVER_INPUT_POLICY_FLAX_FIRST: u8 = 2;

pub fn is_valid_weaver_input_policy(policy: u8) -> bool {
    matches!(
        policy,
        WEAVER_INPUT_POLICY_AUTO | WEAVER_INPUT_POLICY_WOOL_FIRST | WEAVER_INPUT_POLICY_FLAX_FIRST
    )
}

pub fn normalize_weaver_input_policy(policy: u8) -> u8 {
    if is_valid_weaver_input_policy(policy) {
        policy
    } else {
        WEAVER_INPUT_POLICY_AUTO
    }
}

/// Lower ranks win when equal-priority active textile workshops compete for a
/// route-specific cart. Automatic workshops remain a neutral middle pool and
/// the opposite ingredient may still be stored for a later recipe change.
pub fn weaver_fibre_delivery_preference_rank(policy: u8, flax: bool) -> u8 {
    match normalize_weaver_input_policy(policy) {
        WEAVER_INPUT_POLICY_AUTO => 1,
        WEAVER_INPUT_POLICY_WOOL_FIRST if !flax => 0,
        WEAVER_INPUT_POLICY_FLAX_FIRST if flax => 0,
        WEAVER_INPUT_POLICY_WOOL_FIRST | WEAVER_INPUT_POLICY_FLAX_FIRST => 2,
        _ => unreachable!("weaver input policy is normalized"),
    }
}

/// Explicit buttons select one recipe. Auto alone may switch to whichever
/// complete route has the deeper working stock.
pub fn weaver_uses_flax(
    policy: u8,
    wool: f64,
    flax: f64,
    water: f64,
    wool_per_cycle: f64,
    flax_per_cycle: f64,
    water_per_cycle: f64,
) -> bool {
    let wool_cycles = wool.max(0.0) / wool_per_cycle.max(1e-6);
    let flax_cycles =
        (flax.max(0.0) / flax_per_cycle.max(1e-6)).min(water.max(0.0) / water_per_cycle.max(1e-6));
    match normalize_weaver_input_policy(policy) {
        WEAVER_INPUT_POLICY_WOOL_FIRST => false,
        WEAVER_INPUT_POLICY_FLAX_FIRST => true,
        WEAVER_INPUT_POLICY_AUTO => {
            (flax > 1e-6 && wool <= 1e-6) || flax_cycles > wool_cycles + 1e-9
        }
        _ => unreachable!("weaver input policy is normalized"),
    }
}

/// Selects the flax-family route at the Weaver, where retting has already
/// happened and the two complete alternatives are yarn and linen.
pub fn weaver_uses_linen(
    policy: u8,
    yarn: f64,
    linen: f64,
    yarn_per_cycle: f64,
    linen_per_cycle: f64,
) -> bool {
    let yarn_cycles = yarn.max(0.0) / yarn_per_cycle.max(1e-6);
    let linen_cycles = linen.max(0.0) / linen_per_cycle.max(1e-6);
    match normalize_weaver_input_policy(policy) {
        WEAVER_INPUT_POLICY_WOOL_FIRST => false,
        WEAVER_INPUT_POLICY_FLAX_FIRST => true,
        WEAVER_INPUT_POLICY_AUTO => {
            (linen > 1e-6 && yarn <= 1e-6) || linen_cycles > yarn_cycles + 1e-9
        }
        _ => unreachable!("weaver input policy is normalized"),
    }
}

#[cfg(test)]
mod tests {
    use super::{
        is_valid_weaver_input_policy, normalize_weaver_input_policy,
        weaver_fibre_delivery_preference_rank, weaver_uses_flax, weaver_uses_linen,
        WEAVER_INPUT_POLICY_AUTO, WEAVER_INPUT_POLICY_FLAX_FIRST, WEAVER_INPUT_POLICY_WOOL_FIRST,
    };

    fn uses_flax(policy: u8, wool: f64, flax: f64, water: f64) -> bool {
        weaver_uses_flax(policy, wool, flax, water, 3.0, 3.0, 1.0)
    }

    #[test]
    fn old_and_malformed_rows_keep_automatic_selection() {
        assert_eq!(normalize_weaver_input_policy(99), WEAVER_INPUT_POLICY_AUTO);
        assert!(!uses_flax(99, 6.0, 3.0, 1.0));
        assert!(uses_flax(99, 3.0, 6.0, 2.0));
    }

    #[test]
    fn each_preference_uses_its_ready_route_first() {
        assert!(!uses_flax(WEAVER_INPUT_POLICY_WOOL_FIRST, 3.0, 6.0, 2.0));
        assert!(uses_flax(WEAVER_INPUT_POLICY_FLAX_FIRST, 6.0, 3.0, 1.0));
    }

    #[test]
    fn explicit_recipes_do_not_fall_back_to_alternate_stock() {
        assert!(!uses_flax(WEAVER_INPUT_POLICY_WOOL_FIRST, 0.0, 3.0, 1.0));
        assert!(uses_flax(WEAVER_INPUT_POLICY_FLAX_FIRST, 3.0, 3.0, 0.0));
        assert!(uses_flax(WEAVER_INPUT_POLICY_FLAX_FIRST, 0.0, 0.0, 0.0));
    }

    #[test]
    fn the_same_persisted_preference_selects_yarn_or_linen_at_the_weaver() {
        assert!(!weaver_uses_linen(
            WEAVER_INPUT_POLICY_WOOL_FIRST,
            2.0,
            4.0,
            2.0,
            2.0,
        ));
        assert!(weaver_uses_linen(
            WEAVER_INPUT_POLICY_FLAX_FIRST,
            4.0,
            2.0,
            2.0,
            2.0,
        ));
        assert!(!weaver_uses_linen(
            WEAVER_INPUT_POLICY_WOOL_FIRST,
            0.0,
            2.0,
            2.0,
            2.0,
        ));
    }

    #[test]
    fn policy_codes_are_small_and_explicit() {
        for policy in [
            WEAVER_INPUT_POLICY_AUTO,
            WEAVER_INPUT_POLICY_WOOL_FIRST,
            WEAVER_INPUT_POLICY_FLAX_FIRST,
        ] {
            assert!(is_valid_weaver_input_policy(policy));
        }
        assert!(!is_valid_weaver_input_policy(3));
    }

    #[test]
    fn fibre_delivery_ranks_match_specialized_looms_without_disabling_fallback() {
        assert_eq!(
            weaver_fibre_delivery_preference_rank(WEAVER_INPUT_POLICY_WOOL_FIRST, false),
            0
        );
        assert_eq!(
            weaver_fibre_delivery_preference_rank(WEAVER_INPUT_POLICY_FLAX_FIRST, true),
            0
        );
        assert_eq!(
            weaver_fibre_delivery_preference_rank(WEAVER_INPUT_POLICY_AUTO, false),
            1
        );
        assert_eq!(
            weaver_fibre_delivery_preference_rank(WEAVER_INPUT_POLICY_WOOL_FIRST, true),
            2
        );
        assert_eq!(weaver_fibre_delivery_preference_rank(99, true), 1);
    }
}
